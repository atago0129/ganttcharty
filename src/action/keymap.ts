import * as Action from './';

export const keyMap = {
  'up': ['k', 'up'],
  'down': ['j', 'down'],
  'left': ['h', 'left'],
  'right': ['l', 'right'],
  'add': ['enter'],
  'delete': ['backspace']
};

export const handlers = {
  'up': () => {
    Action.Task.selectPrevTask();
  },
  'down': () => {
    Action.Task.selectNextTask();
  },
  'left': () => {
    Action.Option.updateBaseTime(24 * 60 * 60 * 1000);
  },
  'right': () => {
    Action.Option.updateBaseTime(-24 * 60 * 60 * 1000);
  },
  'add': (event?: KeyboardEvent) => {
    event!.preventDefault();
    Action.Task.add();
  },
  'delete': () => {
    Action.Task.deleteSelectedTask();
  }
};
