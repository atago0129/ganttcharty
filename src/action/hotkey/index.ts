import * as Action from '../';
import * as State from '../../state';
import { HotkeysEvent } from 'hotkeys-js';

export const keyMap = {
  'selected up': ['k', 'up'],
  'selected down': ['j', 'down'],
  'move left': ['h', 'left'],
  'move right': ['l', 'right'],
  'add task': ['enter'],
  'remove task': ['backspace'],
  'expand task': ['shift+right'],
  'collapse task': ['shift+left'],
  'toggle export': ['shift+e']
};

export const handlers = {
  'selected up': e => {
    e.preventDefault();
    Action.UI.selectPrevTask();
  },
  'selected down': e => {
    e.preventDefault();
    Action.UI.selectNextTask();
  },
  'move left': e => {
    e.preventDefault();
    const state = State.get()!;
    Action.UI.moveSelectedTask(-State.Option.scaleTime(state.option.scale));
  },
  'move right': e => {
    e.preventDefault();
    const state = State.get()!;
    Action.UI.moveSelectedTask(State.Option.scaleTime(state.option.scale));
  },
  'add task': e => {
    e.preventDefault();
    Action.Task.add(State.get()!.ui.selectedTaskId);
  },
  'remove task': () => {
    const state = State.get()!;
    if (state.ui.selectedTaskId) {
      Action.Task.remove(state.ui.selectedTaskId);
    }
  },
  'expand task': () => {
    const state = State.get()!;
    if (state.ui.selectedTaskId) {
      Action.Task.expand(state.ui.selectedTaskId);
    }
  },
  'collapse task': () => {
    const state = State.get()!;
    if (state.ui.selectedTaskId) {
      Action.Task.collapse(state.ui.selectedTaskId);
    }
  },
  'toggle export': () => {
    Action.UI.toggleExportView(!State.get()!.ui.exporting);
  }
} as {
  [key in keyof typeof keyMap]: (ke: KeyboardEvent, he: HotkeysEvent) => void
};
