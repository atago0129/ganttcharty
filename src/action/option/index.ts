import * as State from '../../state';

export const zoomup = () => {
  State.update(state => {
    state.option.columnWidth += 3;
  });
};

export const zoomdown = () => {
  State.update(state => {
    state.option.columnWidth -= 3;
  });
};
