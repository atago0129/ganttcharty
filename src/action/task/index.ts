import { v4 as uuid } from 'uuid';
import * as State from '../../state';

export const insertPrev = (
  targetTaskId: State.Task.TaskId,
  insertTaskId: State.Task.TaskId
) => {
  State.update(state => {
    const targetTask =
      State.Task.getPrev(state.tasks, targetTaskId)! ||
      State.Task.get(state.tasks, targetTaskId);
    const insertTask = State.Task.get(state.tasks, insertTaskId)!;

    insertTask.parentId = targetTask.parentId;
    state.tasks.splice(state.tasks.indexOf(insertTask), 1);
    state.tasks.splice(state.tasks.indexOf(targetTask), 0, insertTask);
  });
};

export const insertNext = (
  targetTaskId: State.Task.TaskId,
  insertTaskId: State.Task.TaskId
) => {
  State.update(state => {
    const targetTask = State.Task.getNext(state.tasks, targetTaskId)!;
    const insertTask = State.Task.get(state.tasks, insertTaskId)!;

    insertTask.parentId = targetTask.parentId;
    state.tasks.splice(state.tasks.indexOf(insertTask), 1);
    state.tasks.splice(state.tasks.indexOf(targetTask), 0, insertTask);
  });
};

export const expand = (taskId: State.Task.TaskId) => {
  State.update(state => {
    const target = State.Task.get(state.tasks, taskId)!;
    const children = State.Task.getChildren(state.tasks, target.id, false);
    if (target && children.length && target.collapsed) {
      target.collapsed = false;
    }
  });
};

export const collapse = (taskId: State.Task.TaskId) => {
  State.update(state => {
    const target = State.Task.get(state.tasks, taskId)!;
    const children = State.Task.getChildren(state.tasks, target.id, false);
    if (target && children.length && !target.collapsed) {
      target.collapsed = true;
    }
  });
};

export const add = (extendTaskId?: State.Task.TaskId) => {
  State.update(state => {
    if (!state.ui.selectedTaskId) return;

    // create newTask and extends selected task.
    const extend = State.Task.get(state.tasks, extendTaskId)!;
    const newTask: State.Task.Task = {
      id: uuid(),
      name: `new task ${state.tasks.length}`,
      description: '',
      startedAt: new Date(extend.startedAt.getTime()),
      finishedAt: new Date(extend.finishedAt.getTime()),
      parentId: extend.parentId
    };

    // insert newTask to next to selected task.
    state.tasks.splice(state.tasks.indexOf(extend) + 1, 0, newTask);

    // select newTask.
    state.ui.selectedTaskId = newTask.id;
  });
};

export const remove = () => {
  State.update(state => {
    if (!state.ui.selectedTaskId) return;

    // remove selected task.
    const target = State.Task.get(state.tasks, state.ui.selectedTaskId)!;

    // remove children.
    const children = State.Task.getChildren(state.tasks, target.id);
    children.forEach(child =>
      state.tasks.splice(state.tasks.indexOf(child), 1)
    );

    // memory target to move.
    const prev = State.Task.getPrev(state.tasks, state.ui.selectedTaskId);
    const next = State.Task.getNext(state.tasks, state.ui.selectedTaskId);

    // remove target task.
    state.tasks.splice(state.tasks.indexOf(target), 1);

    // move to target.
    if (next) {
      state.ui.selectedTaskId = next.id;
    } else if (prev) {
      state.ui.selectedTaskId = prev.id;
    }

    // clear selecting if remove last task.
    if (state.tasks.length === 0) {
      state.ui.selectedTaskId = undefined;
    }
  });
};

export const update = (
  taskId: State.Task.TaskId,
  attrs: Partial<State.Task.Task>
) => {
  State.update(state => {
    const task = State.Task.get(state.tasks, taskId);
    Object.keys(attrs).forEach(key => {
      // @ts-ignore
      task[key] = attrs[key];
    });
  });
};
