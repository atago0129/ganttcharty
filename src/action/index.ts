import * as Task from './task';
import * as UI from './ui';
import * as Hotkey from './hotkey';
export { Task, UI, Hotkey };

