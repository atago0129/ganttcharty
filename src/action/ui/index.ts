import * as State from '../../state';

export const updateCurrentTime = (diffX: number) => {
  State.update(state => {
    const diffTime =
      State.Option.scaleTime(state.option.scale) *
      (diffX / state.option.columnWidth);
    state.ui.currentTime = new Date(state.ui.currentTime.getTime() + diffTime);
  });
};

export const updateViewport = (rect: { width: number; height: number }) => {
  State.update(state => {
    state.ui.viewportWidth = rect.width;
    state.ui.viewportHeight = rect.height;
  });
};

export const toggleExportView = (toggle: boolean) => {
  State.update(state => {
    state.ui.exporting = toggle;
  });
};

export const select = (taskId?: State.Task.TaskId) => {
  State.update(state => {
    state.ui.selectedTaskId = taskId;
  });
  setTimeout(() => moveToSelectedTask());
};

export const moveToSelectedTask = () => {
  State.update(state => {
    if (!state.ui.selectedTaskId) return;

    const target = State.Task.get(state.tasks, state.ui.selectedTaskId)!;
    const { scale, columnWidth } = state.option;
    const { currentTime, viewportWidth } = state.ui;
    const startTime = currentTime;
    const finishTime = new Date(
      currentTime.getTime() +
        (viewportWidth / columnWidth) * State.Option.scaleTime(scale)
    );

    if (startTime.getTime() > target.finishedAt.getTime()) {
      state.ui.currentTime = new Date(
        target.startedAt.getTime() - State.Option.scaleTime(scale) * 14
      );
    } else if (target.startedAt.getTime() > finishTime.getTime()) {
      state.ui.currentTime = new Date(
        target.startedAt.getTime() - State.Option.scaleTime(scale) * 14
      );
    }
  });
};

export const selectNextTask = () => {
  State.update(state => {
    if (!state.ui.selectedTaskId) return;

    const next = State.Task.getNext(state.tasks, state.ui.selectedTaskId);
    if (next) {
      state.ui.selectedTaskId = next.id;
    }
  });
};

export const selectPrevTask = () => {
  State.update(state => {
    if (!state.ui.selectedTaskId) return;

    const prev = State.Task.getPrev(state.tasks, state.ui.selectedTaskId);
    if (prev) {
      state.ui.selectedTaskId = prev.id;
    }
  });
};
