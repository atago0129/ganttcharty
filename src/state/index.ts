import { define } from 'srimmer';
import * as UI from './ui';
import * as Task from './task';
import * as Option from './option';

export { Task, Option, UI };

export type State = {
  ui: UI.UI;
  option: Option.Option;
  tasks: Task.Task[];
};

export const {
  Provider,
  Consumer,
  update,
  select,
  get
} = define<State>();

