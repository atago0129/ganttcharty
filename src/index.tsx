import React from 'react';
import ReactDOM from 'react-dom';
import * as State from './state';
import { GanttChart } from './component/app/GanttChart';

type Input = {
  option?: Partial<State.Option.Option>;
  ui?: Partial<State.UI.UI>;
  tasks: State.Task.Task[];
};

export { State };

export function render(id: string, state: Input) {
  const element = document.getElementById(id)!;
  ReactDOM.render(
    <State.Provider state={defaults(element, state)}>
      <GanttChart />
    </State.Provider>,
    element
  );
  return {
    get: State.get
  };
}

const defaults = (element: HTMLElement, state: Input): State.State => {
  state.option = State.Option.defaults(state.option || {});
  state.ui = State.UI.defaults(element, state.option, state.ui || {});
  state.tasks = State.Task.defaults(state.tasks || []);
  return state as State.State;
};
