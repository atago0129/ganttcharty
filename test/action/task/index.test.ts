import diff from 'snapshot-diff';
import { fixture, task } from '../../';
import * as State from '../../../src/state';
import * as Action from '../../../src/action';

describe('Action.Task', () => {
  describe('#insertPrev', () => {
    test('insert to children 1', () => {
      State.set(
        fixture([
          task('1', {}),
          task('2', {}),
          task('3', { parentId: '2' }),
          task('4', { parentId: '2' }),
          task('5', {})
        ])
      );

      const state = State.get()!;
      Action.Task.insertNext('2', '4');
      expect(diff(state, State.get())).toMatchSnapshot();
    });
    test('insert to children 2', () => {
      State.set(
        fixture([
          task('1', {}),
          task('2', {}),
          task('3', { parentId: '2' }),
          task('4', { parentId: '2' }),
          task('5', {})
        ])
      );

      const state = State.get()!;
      Action.Task.insertNext('2', '1');
      expect(diff(state, State.get())).toMatchSnapshot();
    });
    test('insert to children 2', () => {
      State.set(
        fixture([
          task('1', {}),
          task('2', {}),
          task('3', { parentId: '2' }),
          task('4', { parentId: '2' }),
          task('5', {})
        ])
      );

      const state = State.get()!;
      Action.Task.insertNext('4', '1');
      expect(diff(state, State.get())).toMatchSnapshot();
    });
  });
});
