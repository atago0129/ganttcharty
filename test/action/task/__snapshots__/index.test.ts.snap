// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action.Task #insertPrev insert to children 1 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -51,19 +51,19 @@
        \\"startedAt\\": 2018-10-22T16:46:32.613Z,
      },
      Object {
        \\"description\\": \\"\\",
        \\"finishedAt\\": 2018-10-23T16:46:32.613Z,
-       \\"id\\": \\"3\\",
+       \\"id\\": \\"4\\",
        \\"name\\": \\"default task name\\",
        \\"parentId\\": \\"2\\",
        \\"startedAt\\": 2018-10-22T16:46:32.613Z,
      },
      Object {
        \\"description\\": \\"\\",
        \\"finishedAt\\": 2018-10-23T16:46:32.613Z,
-       \\"id\\": \\"4\\",
+       \\"id\\": \\"3\\",
        \\"name\\": \\"default task name\\",
        \\"parentId\\": \\"2\\",
        \\"startedAt\\": 2018-10-22T16:46:32.613Z,
      },
      Object {"
`;

exports[`Action.Task #insertPrev insert to children 2 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -37,19 +37,20 @@
    },
    \\"tasks\\": Array [
      Object {
        \\"description\\": \\"\\",
        \\"finishedAt\\": 2018-10-23T16:46:32.630Z,
-       \\"id\\": \\"1\\",
+       \\"id\\": \\"2\\",
        \\"name\\": \\"default task name\\",
        \\"startedAt\\": 2018-10-22T16:46:32.630Z,
      },
      Object {
        \\"description\\": \\"\\",
        \\"finishedAt\\": 2018-10-23T16:46:32.630Z,
-       \\"id\\": \\"2\\",
+       \\"id\\": \\"1\\",
        \\"name\\": \\"default task name\\",
+       \\"parentId\\": \\"2\\",
        \\"startedAt\\": 2018-10-22T16:46:32.630Z,
      },
      Object {
        \\"description\\": \\"\\",
        \\"finishedAt\\": 2018-10-23T16:46:32.630Z,"
`;

exports[`Action.Task #insertPrev insert to children 2 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -37,35 +37,36 @@
    },
    \\"tasks\\": Array [
      Object {
        \\"description\\": \\"\\",
        \\"finishedAt\\": 2018-10-23T16:46:32.634Z,
-       \\"id\\": \\"1\\",
+       \\"id\\": \\"2\\",
        \\"name\\": \\"default task name\\",
        \\"startedAt\\": 2018-10-22T16:46:32.634Z,
      },
      Object {
        \\"description\\": \\"\\",
        \\"finishedAt\\": 2018-10-23T16:46:32.634Z,
-       \\"id\\": \\"2\\",
+       \\"id\\": \\"3\\",
        \\"name\\": \\"default task name\\",
+       \\"parentId\\": \\"2\\",
        \\"startedAt\\": 2018-10-22T16:46:32.634Z,
      },
      Object {
        \\"description\\": \\"\\",
        \\"finishedAt\\": 2018-10-23T16:46:32.634Z,
-       \\"id\\": \\"3\\",
+       \\"id\\": \\"4\\",
        \\"name\\": \\"default task name\\",
        \\"parentId\\": \\"2\\",
        \\"startedAt\\": 2018-10-22T16:46:32.634Z,
      },
      Object {
        \\"description\\": \\"\\",
        \\"finishedAt\\": 2018-10-23T16:46:32.634Z,
-       \\"id\\": \\"4\\",
+       \\"id\\": \\"1\\",
        \\"name\\": \\"default task name\\",
-       \\"parentId\\": \\"2\\",
+       \\"parentId\\": undefined,
        \\"startedAt\\": 2018-10-22T16:46:32.634Z,
      },
      Object {
        \\"description\\": \\"\\",
        \\"finishedAt\\": 2018-10-23T16:46:32.634Z,"
`;
